packages/workflow/src/records/handler/download.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * OpenCRVS is also distributed under the terms of the Civil Registration
+ * & Healthcare Disclaimer located at http://opencrvs.org/license.
+ *
+ * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
+ */
+
+import {
+  getStatusFromTask,
+  getTaskFromSavedBundle,
+  hasScope,
+  inScope,
+  TaskStatus,
+  ValidRecord
+} from '@opencrvs/commons/types'
+import * as Hapi from '@hapi/hapi'
+import * as z from 'zod'
+import { validateRequest } from '@workflow/utils/index'
+import { getRecordById } from '@workflow/records/index'
+import { getToken, isSystem } from '@workflow/utils/authUtils'
+import { IAuthHeader } from '@opencrvs/commons'
+import { toDownloaded } from '@workflow/records/state-transitions'
+import { getLoggedInPractitionerResource } from '@workflow/features/user/utils'
+
+function getDownloadedOrAssignedExtension(
+  authHeader: IAuthHeader,
+  status: TaskStatus
+) {
+  if (
+    inScope(authHeader, ['declare', 'recordsearch']) ||
+    (hasScope(authHeader, 'validate') && status === 'VALIDATED')
+  ) {
+    return `http://opencrvs.org/specs/extension/regDownloaded` as const
+  }
+  return `http://opencrvs.org/specs/extension/regAssigned` as const
+}
+
+export async function downloadRecordHandler(
+  request: Hapi.Request,
+  h: Hapi.ResponseToolkit
+) {
+  const payload = validateRequest(
+    z.object({
+      id: z.string()
+    }),
+    request.payload
+  )
+
+  const token = getToken(request)
+  const record = await getRecordById(
+    // Task history is fetched rather than the task only
+    `${payload.id}?includeHistoryResources`,
+    token,
+    [
+      'CERTIFIED',
+      'VALIDATED',
+      'IN_PROGRESS',
+      'READY_FOR_REVIEW',
+      'REGISTERED',
+      'ISSUED',
+      'CORRECTION_REQUESTED'
+    ]
+  )
+
+  const task = getTaskFromSavedBundle(record)
+  const businessStatus = getStatusFromTask(task)
+
+  if (!businessStatus) {
+    throw new Error("Task didn't have any status. This should never happen")
+  }
+
+  const practitioner = await getLoggedInPractitionerResource(token)
+
+  const extensionUrl = getDownloadedOrAssignedExtension(
+    request.headers.authHeader,
+    businessStatus
+  )
+
+  const downloadedRecord = await toDownloaded(
+    token,
+    record,
+    isSystem(request),
+    practitioner,
+    extensionUrl
+  )
```

Comment:
How long does it take to execute the download handler? I'm slightly worries that making

1. A fetch to Hearth inside `getLoggedInPractitionerResource` to fetch the practitioner
2. Inside `toDownloaded -> setupLastRegLocation` (packages/workflow/src/features/registration/fhir/fhir-bundle-modifier.ts:474)
  2.1 Fetching practitionerRole in `packages/workflow/src/features/user/utils.ts:171`
    2.1.1 For each practitioner location fetching location from Hearth
  2.2 Fetching practitioner office `packages/workflow/src/features/registration/fhir/fhir-bundle-modifier.ts:494`
    2.2.1 For each practitioner location fetching location from Hearth
3. Inside `toDownloaded` sending a request to Hearth and to Search API

Good rule of thumb is that a request to Hearth take one second as it's **extremely** slow. Even plain HTTP requests are very resource and time intensive as the protocol involves a lot of handshaking and verifications. That's one of the reasons why we now fetch the bundle directly from MongoDB bypassing Hearth. So simple calculation based on this is, it can take 7-8s to finish the operation. As we are only making requests to our local database, the reality is it shouldn't take more than 100ms, 1 second as most. 

Can we find an alternative way to implement this? I think it was @Zangetsu101 who proposed the FHIR bundle aggregator would also return facts about the practitioner making the query which I think would solve the majority of your issue here. The user downloading a record doesn't need to wait until their downloaded action is stored to Hearth either and we could give them the response even before that is done.

Can we refactor these parts of the code so that there would be as little HTTP

Reply:
https://github.com/opencrvs/opencrvs-core/pull/6360/files#diff-f9112c4dac6b5d4e553d2894b95b40e6d1d7c75325ff857c42a87460d67445b6R272

Not waiting for `indexBundle` into search apparently saves a lot of time. I've kept it like this for now. However, this could be discussed for a better understanding.

Reply:
Yeah, that's true. Did not keep it like this from the start. We are only saving the task in hearth and search. But we're sending the whole record bundle back. That's why it's like this now. Forgot its name and what it is actually supposed to do.